src/Processors/QueryPlan/Optimizations/QueryPlanOptimizationSettings.h

```diff
@@ -83,6 +88,8 @@ struct QueryPlanOptimizationSettings
 
     bool build_sets = true;
 
+    bool keep_logical_steps = false;
```

Comment:
I do not see when it could be `true`.

Reply:
It's set from `EXPLAIN` options, not query settings (since it makes sense only in explain)
For example `EXPLAIN PLAN actions = 1, optimize = 1, logical = 1 SELECT ...`

https://github.com/ClickHouse/ClickHouse/blob/bff7904498d53f06bf709e8ed3156acd6f6e4b43/src/Processors/QueryPlan/Optimizations/QueryPlanOptimizationSettings.cpp#L75-L76

https://github.com/ClickHouse/ClickHouse/blob/bff7904498d53f06bf709e8ed3156acd6f6e4b43/src/Interpreters/InterpreterExplainQuery.cpp#L519-L523